typed-racket-lib/typed-racket/private/type-contract.rkt

```diff
@@ -683,7 +716,30 @@
             (fail #:reason (~a "cannot import structure types from"
                                "untyped code"))
             (struct-type/sc null))]
-       [(Struct-Property: s) (struct-property/sc (t->sc s))]
+       [(Struct-Property: s _) (struct-property/sc (t->sc s))]
+       [(Has-Struct-Property: orig-id)
+        ;; we can't call syntax-local-value/immediate in has-struct-property case in parse-type
+        (define-values (a prop-name) (syntax-local-value/immediate orig-id (λ () (values #t orig-id))))
+        (match-define (Struct-Property: _ pred?) (lookup-id-type/lexical prop-name))
+        ;; if original-name is only set when the type is added via require/typed
+        
+        ;; the original-name of `prop-name` is its original referece in the unexpanded program.
+        (define real-prop-var (or (syntax-property prop-name 'original-name) prop-name))
+
+        ;; a property is wrapped so we need its original reference
+        (define real-pred-var (or (syntax-property pred? 'original-name) (syntax-e pred?)))
+        
+        ;; the `pred?` could be provided to a property through require/typed, 
+        ;; so we need to check if it is produced by the property
+        (flat/sc #`(if (not (struct-type-property-predicate-procedure? #,pred? #,real-prop-var))
+                       (raise-arguments-error 'struct-property
```

Comment:
This error should probably go under a lambda, so it's not a bother unless the code actually uses the contract.

Reply:
like `(lambda _ (raise-arguments-error ...))` ? If so, we would need wrap the lambda into a contract because flat/sc only accepts a contract. But that would look a little weird to me, since I don't usually see a contract always raise an errro.

My intention was to fail fast when an unmatched predicate was given, but delaying the error looks better for this code. 

So I think maybe the whole if expression should go under a lambda, which we turn into a flat-named-contract